flink-clients/src/main/java/org/apache/flink/client/program/ContextEnvironment.java

```diff
@@ -165,12 +181,65 @@ public ContextEnvironmentFactory(Client client, List<URL> jarFilesToAttach,
 		
 		@Override
 		public ExecutionEnvironment createExecutionEnvironment() {
-			ContextEnvironment env = new ContextEnvironment(client, jarFilesToAttach, classpathsToAttach,
-					userCodeClassLoader, wait);
-			if (defaultParallelism > 0) {
-				env.setParallelism(defaultParallelism);
+			if (wait || lastEnv == null) {
```

Comment:
Could you at least add a comment here? This condition is not very intuitive. Why do you even set `lastEnv` here for blocking execution mode?